tokio/src/task/scope/intrusive_double_linked_list.rs

```diff
@@ -0,0 +1,628 @@
+//! An intrusive double linked list of data
+
+use core::marker::PhantomPinned;
+use core::ops::{Deref, DerefMut};
+use core::ptr::null_mut;
+
+/// A node which carries data of type `T` and is stored in an intrusive list
+#[derive(Debug)]
+pub(crate) struct ListNode<T> {
+    /// The previous node in the list. null if there is no previous node.
+    prev: *mut ListNode<T>,
+    /// The next node in the list. null if there is no previous node.
+    next: *mut ListNode<T>,
```

Comment:
I would strongly advocate for the use of `Option<ptr::NonNull<...>>` throughout this module...

Reply:
I guess I could try if it really helps. Sometimes unwrapping from `Option`s is quite unergonomic. But you called out some good examples where it would.

I might also do that change in futures-intrusive first.